bulk/bulk.go

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2025 Greptime Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bulk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	gpb "github.com/GreptimeTeam/greptime-proto/go/greptime/v1"
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table/types"
+	"github.com/apache/arrow/go/v17/arrow/flight"
+	"github.com/apache/arrow/go/v17/arrow/ipc"
+	"google.golang.org/grpc"
+
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table"
+)
+
+type BulkClient struct {
+	client flight.FlightServiceClient
+}
+
+func NewBulkClient(conn *grpc.ClientConn) *BulkClient {
+	return &BulkClient{
+		client: flight.NewFlightServiceClient(conn),
+	}
+}
+
+type BulkWriter interface {
+	Write(table *table.Table) error
+	Close() error
+}
+
+type bulkWriter struct {
+	client *BulkClient
+	stream flight.FlightService_DoPutClient
+	writer *flight.Writer
+	mu     sync.Mutex
```

Comment:
Is it necessary to use the `mu`? What kind of critical resources does it protect?